src/create_disk.sh

```diff
@@ -224,6 +191,11 @@ case "$arch" in
         # shellcheck disable=SC2206
         sgdisk_args+=(-n ${BOOTPN}:0:+384M -c ${BOOTPN}:boot \
                       -n ${ROOTPN}:0:"${rootfs_size}" -c ${ROOTPN}:root -t ${ROOTPN}:0FC63DAF-8483-4772-8E79-3D69D8477DE4)
+        if [[ ${secure_execution} -eq 1 ]]; then
+            # shellcheck disable=SC2206
+            sgdisk_args+=(-n ${BOOTVERITYHASHPN}:0:+128M -c ${BOOTVERITYHASHPN}:boothash \
+                          -n ${ROOTVERITYHASHPN}:0:+256M -c ${ROOTVERITYHASHPN}:roothash)
```

Comment:
Wait, where is the additional space coming from?  `cmd-buildextend-metal` creates the containing disk without factoring these in.

Also, the sizes don't seem right.  In my tests, with a 512-byte blocksize, a 384 MiB bootfs needs 25 MiB in hashes, and a 16 GiB rootfs needs 1.1 GiB.  The extra padding in the former will compress out, of course, but I don't see how the latter can work.

I'd be tempted to suggest writing the hashes to files first, then creating the partitions based on the file sizes.  That's awkward, though, when the disk sizing happens in a different script.

Reply:
I guess there no additional space is required, by default `rootfs` is rather small, and gets grown to 16Gb during firstboot. Here some info from build on s390x:
```
Number  Start (sector)    End (sector)  Size       Code  Name
   1            2048          411647   200.0 MiB   8300  se
   3          411648         1198079   384.0 MiB   8300  boot
   4         1198080         5380095   2.0 GiB     8300  root
   5         5380096         5642239   128.0 MiB   8300  boothash
   6         5642240         6166527   256.0 MiB   8300  roothash

+ lsblk -f
NAME     FSTYPE         FSVER LABEL UUID                                 FSAVAIL FSUSE% MOUNTPOINTS
vdb                                                                                     
|-vdb3   ext4           1.0   boot  96d15588-3596-4b3c-adca-a2ff7279ea63                
| `-boot ext4           1.0   boot  96d15588-3596-4b3c-adca-a2ff7279ea63  218.4M    31% /tmp/rootfs/boot
|-vdb4   xfs                  root  910678ff-f77e-4a7d-8d53-86f2ac47a823                
| `-root xfs                  root  910678ff-f77e-4a7d-8d53-86f2ac47a823    576M    72% /tmp/rootfs
```
And sizes for hashes:
```
-rw-------. 1 root root  25M Aug 26 11:26 /tmp/boothash
-rw-------. 1 root root 129M Aug 26 11:26 /tmp/roothash
```
So that sizes of both hashes fit well (even when i `dd if=/dev/random of=$rootfs/garbage bs=128M count=2` to make `FSUSE` about 90%)

Reply:
That's weird, @nikita-dubrovskii. I'm not sure how the root partition is only 2 GiB in your setup; that's drastically lower than at least x86_64. Do you have the output of `cosa buildextend-secex` where it says "Disk size estimated to"?

> Wait, where is the additional space coming from? `cmd-buildextend-metal` creates the containing disk without factoring these in.

16 GiB is the size of the image, not the rootfs. Here we're setting up dm-verity only on the bootfs and rootfs partitions. The size for the latter comes from `src/estimate-commit-disk-size`, which calculates it from the OSTree commit and then `cmd-buildextend-metal` tells it to add 35%. Right now, the calculated rootfs for RHCOS weighs in at ~3.5 GiB, so there's ample space to add hash partitions after the rootfs.

> Also, the sizes don't seem right. In my tests, with a 512-byte blocksize, a 384 MiB bootfs needs 25 MiB in hashes, and a 16 GiB rootfs needs 1.1 GiB. The extra padding in the former will compress out, of course, but I don't see how the latter can work.

So using the 3.5 GiB figure, I get a hashfile size of ~221M. This fits within the current roothash partition size, but indeed it's cutting it close and if RHCOS gets larger in the future, we'd run out. I do like your dynamic sizing idea, and I don't think it'd be that hacky since it's all within `create_disk.sh`. I'll try to tackle that, though OTOH it seems like the current sizes should work for now so I'm inclined to get this in and do it in a follow-up. WDYT?

Reply:
Yup, here is the output for `buidextend-secex`
```
Disk size estimated to 2555M
```

Reply:
Comment above is valid only for FCOS, here is RHCOS:
```
Disk size estimated to 4176M
```

Reply:
Yup, sorry, I missed the `image_size` dance in `cmd-buildextend-metal`.  Should we add a comment in `create_disk.sh` saying that `secex` assumes a qcow, which includes extra space for additional partitions?

I'm okay deferring optimizations to a follow-up, sure.  (And I still think there's value to using a 4K block size to reduce the incompressible data by a factor of 8.)

Reply:
Disk layout and hashes on s390x RHCOS build:
```
Number  Start (sector)    End (sector)  Size       Code  Name
   1            2048          411647   200.0 MiB   8300  se
   3          411648         1198079   384.0 MiB   8300  boot
   4         1198080         8699903   3.6 GiB     8300  root
   5         8699904         8962047   128.0 MiB   8300  boothash
   6         8962048         9486335   256.0 MiB   8300  roothash
```

```
-rw-------. 1 root root 231M Aug 26 19:16 /tmp/roothash
```

Reply:
> Yup, sorry, I missed the `image_size` dance in `cmd-buildextend-metal`. Should we add a comment in `create_disk.sh` saying that `secex` assumes a qcow, which includes extra space for additional partitions?

Good point, will do as part of the follow-up!